odoo/addons/base/models/res_partner.py

```diff
@@ -979,9 +979,12 @@ def address_get(self, adr_pref=None):
                         result[record.type] = record.id
                     if len(result) == len(adr_pref):
                         return result
-                    to_scan = [c for c in record.child_ids
-                                 if c not in visited
-                                 if not c.is_company] + to_scan
+                    domain = [("id", "not in", [i.id for i in visited]),
+                              ("parent_id", "=", record.id), ("is_company", "!=", True)]
+                    for address_type in adr_pref - set(result):
+                        new_address = self.search(domain + [("type", "=", address_type)], limit=1)
```

Comment:
Couldn't be possible to use `read_group` here grouping by address type?

That way, you will retrieve all the info by running only one query

Reply:
I have commented in the description about this case using `read_group`

Reply:
I mean, if you use read_group you will get the 555k records using an aggregator array_agg

However, the odoo API is not supporting the aggregator `FIRST(id order by %s) % self._name`

Another way is to use directly SQL, but I prefer using search(..., limit=1) instead

Let me know if you have the diff that avoid retrieving all the children

Reply:
Also, consider the method is only called with only 2 preferred address

So, using a small partner with only 3 children it is even faster

Reply:
> However, the odoo API is not supporting the aggregator `FIRST(id order by %s) % self._name`

You're right!

However, if you don't care about order, you could use:

`min(id)`

Reply:
Yes, it was considered `max` or `min` will retrieve a different value than the current one returned

Maybe for master is good but for stable is better to return the same record and the order matters here

I mean, `self.env["res.partner"]._order` returns `'display_name ASC, id DESC'`

So, when you use `partner.child_ids` they are analyzed in the same order

Notice the following query:

```sql
SELECT "res_partner".id
FROM "res_partner"
WHERE (("res_partner"."active" = true)
    AND ("res_partner"."parent_id" in (2)))
ORDER BY  "res_partner"."display_name" ASC, "res_partner"."id" DESC
```

But the benefit is only `30ms` so not big deal using read_group instead comparing with returning different values in stable

Reply:
Got it, thanks for considering it.

Reply:
Thank you for forcing to me to add good feedback that I missed in the main description